### ISSUE TITLE
[Feature] Implement Computer Vision (CV) Model Failure Analysis

### ISSUE THREAD

**Issue by Shriyabh11** [etsi-ai/etsi-failprint#38]
# [Feature] Implement Computer Vision (CV) Model Failure Analysis

## Description

Implement computer vision (CV) model failure analysis capabilities to extend **failprint** beyond structured and text data. This will enable automatic root cause analysis for tasks such as image classification, object detection, and image segmentation.

This issue is the second part of the "Support for unstructured data" epic, following the successful implementation of NLP failure analysis.

---

## Motivation

Just like NLP models, computer vision models can fail in systematic but hard-to-diagnose ways. Visual patterns—such as low-light conditions, specific object categories, image corruption, or domain shifts—can cause a model to fail consistently. Manually sifting through thousands of failed images is impractical.

This feature aims to automate the surfacing of these patterns, making **failprint** an invaluable tool for CV practitioners and completing its MLOps utility for all major data modalities.

---

### 1. Create Image Preprocessing Pipeline

- Develop a standard pipeline to load, resize, and normalize images to prepare them for embedding models.

### 2. Implement Embedding-Based Clustering

- Integrate pre-trained CV models (e.g., ResNet, ViT, CLIP from `torchvision` or `timm`) to generate embeddings for failed images.
- Use clustering algorithms (e.g., DBSCAN, K-Means) on these embeddings to group visually similar failures together.

### 3. Create CV-Focused Failure Segmentation

- Extract statistical features from each image:
  - Brightness
  - Contrast
  - Color distributions
  - Aspect ratio
  - Image resolution
- Use the existing `segment_failures` function to find which visual properties are over-represented in the failure set.

### 4. Integrate with Popular CV Libraries

- Ensure smooth integration with libraries such as Pillow, OpenCV, and torchvision for data loading and processing.

### 5. (Stretch Goal) Add Image-Specific Drift Detection

- Implement methods to detect drift between the training image set and the failure set, looking for shifts in brightness, color, or the overall embedding space.

---

## Example Usage

```python
import pandas as pd
from etsi.failprint import analyze_cv # New proposed function

# List of paths to the image files
image_paths = [
    "data/images/cat_01.jpg", 
    "data/images/cat_02_dark.jpg", 
    "data/images/dog_01.jpg", 
    "data/images/dog_02_blurry.jpg",
    "data/images/cat_03.jpg"
]

# 0 for 'cat', 1 for 'dog'
y_true = [0, 0, 1, 1, 0]
y_pred = [0, 1, 1, 0, 0] # Model fails on a dark cat and a blurry dog

# Analyze CV model failures
report = analyze_cv(
    image_paths=image_paths,
    y_true=y_true,
    y_pred=y_pred,
    model_name="image_classifier",
    embedding_model="resnet50", # Allow user to choose an embedding model
    cluster_failures=True,
    output="markdown"
)

print(report)
```

---

### 1. Visual Failure Clustering

- The report should display clusters of visually similar images that the model failed on (e.g., a collage of low-light images, a group of blurry images).

### 2. Image Property Analysis

- Statistical breakdown showing if failures are correlated with specific attributes, such as:
  - Mean brightness or contrast
  - Image dimensions or aspect ratio
  - Dominant color analysis

### 3. Drift Detection

- Insights into how the visual characteristics of the failure set differ from a baseline (training) set.

### 4. Clear Reporting

- The final markdown report should be easy to interpret, with image thumbnails or collages for each cluster to provide instant visual feedback.

---

**Related Epic**: Support for unstructured data  
**Preceding Issue**: NLP Failure Analysis Implementation

---
